eidas-saml-parser/src/main/java/uk/gov/ida/notification/eidassaml/EidasSamlResource.java

```diff
@@ -29,44 +39,55 @@
 public class EidasSamlResource {
 
-    private EidasAuthnRequestValidator eidasAuthnRequestValidator;
-    private SamlRequestSignatureValidator<AuthnRequest> samlRequestSignatureValidator;
-    private String x509EncryptionCertString;
-    private String destination;
+    private final EidasAuthnRequestValidator eidasAuthnRequestValidator;
+    private final MetatronProxy metatronProxy;
 
     public EidasSamlResource(EidasAuthnRequestValidator eidasAuthnRequestValidator,
-                             SamlRequestSignatureValidator<AuthnRequest> samlRequestSignatureValidator,
-                             String x509EncryptionCertString,
-                             String destination) {
+                             MetatronProxy metatronProxy) {
         this.eidasAuthnRequestValidator = eidasAuthnRequestValidator;
-        this.samlRequestSignatureValidator = samlRequestSignatureValidator;
-        this.x509EncryptionCertString = x509EncryptionCertString;
-        this.destination = destination;
+        this.metatronProxy = metatronProxy;
     }
 
     @POST
     @Valid
-    public EidasSamlParserResponse post(@Valid EidasSamlParserRequest request) throws UnmarshallingException, XMLParserException {
+    public EidasSamlParserResponse post(@Valid EidasSamlParserRequest request) throws UnmarshallingException, XMLParserException, CertificateException, SignatureException {
 
-        AuthnRequest authnRequest = ObjectUtils.unmarshall(
-                new ByteArrayInputStream(Base64.getDecoder().decode(request.getAuthnRequest().getBytes())),
-                AuthnRequest.class);
+        AuthnRequest authnRequest = unmarshallRequest(request);
+
+        CountryMetadataResponse metatronResponse = getMetatronResponse(authnRequest);
 
-        samlRequestSignatureValidator.validate(authnRequest, SPSSODescriptor.DEFAULT_ELEMENT_NAME);
+        this.isRequestSignatureValid(authnRequest, this.getSigningCredential(metatronResponse.getSamlSigningCertX509()));
         eidasAuthnRequestValidator.validate(authnRequest);
 
+        this.doLogging(authnRequest);
+
+        return new EidasSamlParserResponse(
+                authnRequest.getID(),
+                authnRequest.getIssuer().getValue(),
+                metatronResponse.getSamlEncryptionCertX509(),
+                metatronResponse.getDestination().toString());
+    }
+
+    private AuthnRequest unmarshallRequest(EidasSamlParserRequest request) throws UnmarshallingException, XMLParserException {
+        String xml = new String(Base64.getDecoder().decode(request.getAuthnRequest().getBytes()));
+        System.out.println(xml);
+        return ObjectUtils.unmarshall(
+                new ByteArrayInputStream(xml.getBytes()),
+                AuthnRequest.class);
+    }
+
+    private CountryMetadataResponse getMetatronResponse(AuthnRequest authnRequest) {
+        this.validateIssuer(authnRequest);
+        return metatronProxy.getCountryMetadata(authnRequest.getIssuer().getValue());
```

Comment:
after this line we should check that authnRequest issuer matches CountryMetadataResponse.entityid

Reply:
☝️ John and I talked this through, and decided this check would add no extra value.